networkx/algorithms/cluster.py

```diff
@@ -247,6 +247,72 @@ def wt(u, v):
         yield (i, dtotal, dbidirectional, float(directed_triangles))
 
 
+@not_implemented_for("directed")
+@nx._dispatchable
+def all_triangles(G, nodes=None):
+    """
+    Yields all unique triangles in an undirected graph using fixed node ordering.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        An undirected graph. Self-loops and multiple edges are ignored.
+
+    nodes : node, iterable of nodes, or None (default=None)
+        If a node or iterable of nodes is given, only triangles involving those nodes are yielded.
+        If None, yields all unique triangles in the graph.
+
+    Yields
+    -------
+    tuple
+        A tuple of three nodes forming a triangle (u, v, w) with u < v < w.
+
+    Examples
+    --------
+    >>> G = nx.complete_graph(4)
+    >>> list(all_triangles(G))
+    [(0, 1, 2), (0, 1, 3), (0, 2, 3), (1, 2, 3)]
+    """
+    # If nodes is None, then yield all the triangles of the graph
+    if nodes is None:
+        # Pre-computation based on fixed ID ordering
+        neighbor_sets = {
+            node: {nbr for nbr in G.neighbors(node) if nbr > node} for node in G
```

Comment:
This optimization is invalid in networkx - it is never guaranteed that nodes are ordered this way. In fact, it's not even guaranteed that nodes are comparable!